skimage/morphology/_watershed.pyx

```diff
@@ -10,87 +10,115 @@ All rights reserved.
+        output, or the algorithm will never terminate and blow up your memory!
+    structure : array of int
+        A list of coordinate offsets to compute the raveled coordinates of each
+        neighbor from the raveled coordinates of the current pixel.
+    mask : array of int
+        An array of the same shape as `image` where each pixel contains a
+        nonzero value if it is to be considered for flooding with watershed,
+        zero otherwise. NOTE: it is *essential* that the border pixels (those
+        with neighbors falling outside the volume) are all set to zero, or
+        segfaults could occur.
+    strides : array of int
+        An array representing the number of steps to move along each dimension.
+        This is used in computing the Euclidean distance between raveled
+        indices.
+    compactness : float
+        A value >0 implements the compact watershed algorithm (see .py file).
+    output : array of int
+        The output array, which must already contain nonzero entries at all the
+        seed locations.
     """
     cdef Heapitem elem
     cdef Heapitem new_elem
     cdef Py_ssize_t nneighbors = structure.shape[0]
     cdef Py_ssize_t i = 0
+    cdef Py_ssize_t age = 1
     cdef Py_ssize_t index = 0
-    cdef Py_ssize_t old_index = 0
-    cdef Py_ssize_t max_index = image.shape[0]
 
     cdef Heap *hp = <Heap *> heap_from_numpy2()
 
-    for i in range(pq.shape[0]):
-        elem.value = pq[i, 0]
-        elem.age = pq[i, 1]
-        elem.index = pq[i, 2]
+    for i in range(marker_locations.shape[0]):
+        index = marker_locations[i]
+        elem.value = image[index]
+        elem.age = 0
+        elem.index = index
+        elem.source = index
         heappush(hp, &elem)
 
     while hp.items > 0:
-        #
-        # Pop off an item to work on
-        #
         heappop(hp, &elem)
-        ####################################################
-        # loop through each of the structuring elements
-        #
-        old_index = elem.index
+        if output[elem.index] and elem.index != elem.source:
+            # non-marker, already visited from another neighbor
+            continue
+        output[elem.index] = output[elem.source]
         for i in range(nneighbors):
             # get the flattened address of the neighbor
-            index = structure[i, 0] + old_index
-            if index < 0 or index >= max_index or output[index] or \
-                    not mask[index]:
+            index = structure[i] + elem.index
+            if output[index] or not mask[index]:
+                # previously visited, masked, or border pixel
                 continue
 
             age += 1
             new_elem.value = image[index]
+            if compactness > 0:
+                new_elem.value += (compactness *
+                                   _euclid_dist(index, elem.source, strides))
             new_elem.age = age
```

Comment:
Since you are here adding the "compact watershed", I would also consider adding the geodesic watershed:

```python
            if gradient_cost > 0 or euclidean_cost > 0:
                new_elem.value += sqrt((gradient_cost*(new_elem.value - elem.value)**2 +
                                       euclidean_cost * _euclid_dist(index, elem.index, strides)**2))
```

![figure_1](https://cloud.githubusercontent.com/assets/802153/17175931/3648dc5c-5402-11e6-9284-e0b933228c57.png)

Here I used:

```python
    cdef double euclidean_cost = 0.001
    cdef double gradient_cost = 1.0
```

but these should be passed as parameters.

Reference paper:

T. Jones, A. Carpenter and P. Golland,
"Voronoi-Based Segmentation of Cells on Image Manifolds"
CVBIA05 (535-543), 2005

Example implementations (both from CellProfiler):
https://github.com/kevin-keraudren/cell-segmentation/blob/master/dev/EBImage/src/propagate.cpp
https://github.com/CellProfiler/centrosome/blob/master/centrosome/_propagate.pyx